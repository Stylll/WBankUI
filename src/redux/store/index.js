import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';
import axios from 'axios';

import rootReducer from '../reducers';
import rootSaga from '../sagas';
import { jwtKey } from '../../helpers/defaults';

const sagaMiddleware = createSagaMiddleware();
// const reduxRouterMiddleware = routerMiddleware(history);
const middleware = composeWithDevTools(applyMiddleware(sagaMiddleware));
const store = createStore(rootReducer, middleware);

// eslint-disable-next-line no-undef
const token = localStorage.getItem(jwtKey);

axios.defaults.headers.common.authorization = token;

sagaMiddleware.run(rootSaga);

export default store;
