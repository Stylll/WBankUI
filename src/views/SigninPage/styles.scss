
@import '../../partials/colors';

.container.signinpage {
    display: flex;
    justify-content: center;
    text-align: center;

    .form-container {
        width: 550px;
    }

    .signin-container {
        box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
        margin: 40px;
        padding: 40px;

        .link {
            font-weight: 100;
            &:hover {
                text-decoration: underline;
                cursor: pointer;
            }
        }
        .error-text {
            color: $red;
        }
    }
}

/* small to medium media query */
@media only screen and (max-width: 600px) {
    .container.signinpage {
        .form-container {
            width: auto;
        }
    }
}
