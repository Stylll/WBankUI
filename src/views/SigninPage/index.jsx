import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';
import Header from '../../components/Header/Header.component';
import TextInput from '../../components/TextInput/TextInput.component';
import Button from '../../components/Button/Button.component';
import { validateSignin } from '../../helpers/validators/authValidator';
import * as userActions from '../../redux/actionCreators/userActions';
import './styles.scss';

const SigninPage = ({
  signinUser, isLoading: reduxIsLoading, userErrorMessage,
}) => {
  const [values, setValues] = useState({
    customerId: '',
    email: '',
  });

  const [formErrors, setFormErrors] = useState({
    customerId: '',
    email: '',
  });

  const [isLoading, setIsLoading] = useState(false);

  const history = useHistory();

  useEffect(() => {
    if (!reduxIsLoading) {
      setIsLoading(false);
    }
  }, [reduxIsLoading]);

  const handleChange = (event) => {
    event.preventDefault();
    setValues({
      ...values,
      [event.target.name]: event.target.value,
    });
  };

  const handleClick = (event) => {
    event.preventDefault();
    const { isValid, errors } = validateSignin(values);
    setFormErrors({
      ...errors,
    });
    if (!isValid) {
      return;
    }

    setIsLoading(true);

    signinUser({
      customerId: values.customerId,
      email: values.email,
    });
  };

  const navigateToSignup = () => {
    history.push('/signup');
  };

  const renderSigninForm = (errorMessage) => (
        <div className="signin-container">
            <div>
                <p>Sign in to account</p>
                <span className="error-text">{errorMessage}</span>
                <TextInput
                    name="customerId"
                    placeholder="3214"
                    title="Enter your customer id"
                    type="text"
                    value={values.customerId}
                    error={formErrors.customerId}
                    onChange={handleChange} />
                <TextInput
                    name="email"
                    placeholder="jimhalpert@dundermuflin.com"
                    title="Enter your email"
                    type="text"
                    value={values.email}
                    error={formErrors.email}
                    onChange={handleChange} />
                <Button
                    title="Signin"
                    handleClick={handleClick}
                    showLoader={isLoading}
                    disabled={isLoading}
                />
            </div>
            <br />
            <h4 className="link" onClick={() => { navigateToSignup(); }}>
              Don't have an account?. click here to signup
            </h4>
        </div>
  );

  return (
        <div className="">
            <Header />
            <div className="signinpage container">
                <div className="form-container">
                    <h2 className="intro-text">Welcome To WBank</h2>
                    {renderSigninForm(userErrorMessage)}
                </div>
            </div>
        </div>
  );
};

SigninPage.propTypes = {
  signinUser: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  userErrorMessage: PropTypes.string.isRequired,
};

const mapStateToProps = ({ user }) => ({
  isLoading: user.isLoading,
  userErrorMessage: user.errorMessage,
});

const mapDispatchToProps = {
  signinUser: userActions.signinUser,
};

export default connect(mapStateToProps, mapDispatchToProps)(SigninPage);
