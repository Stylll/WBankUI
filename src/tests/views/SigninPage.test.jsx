import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import React from 'react';

import SigninPage from '../../views/SigninPage/index';
import store from '../../redux/store/index';

describe('SigninPage Test', () => {
  it('should render properly', () => {
    const { baseElement } = render(
        <Provider store={store}>
            <SigninPage />
        </Provider>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should validate signin form', () => {
    const { baseElement, getByTestId, getByText } = render(
        <Provider store={store}>
            <SigninPage />
        </Provider>,
    );
    fireEvent.click(getByTestId('signin-signin'));
    expect(getByText('customerId cannot be empty')).toBeInTheDocument();
    expect(getByText('email cannot be empty')).toBeInTheDocument();
  });

  it('should require valid customer Id', () => {
    const { baseElement, getByTestId, getByText } = render(
        <Provider store={store}>
            <SigninPage />
        </Provider>,
    );
    fireEvent.change(getByTestId('signin-customerId'),
      { target: { value: 'customer', name: 'customerId' } });
    fireEvent.click(getByTestId('signin-signin'));
    expect(getByText('customerId must be a number')).toBeInTheDocument();
    expect(getByText('email cannot be empty')).toBeInTheDocument();
  });

  it('should update the isloading state', () => {
    const { baseElement, getByTestId } = render(
        <Provider store={store}>
            <SigninPage />
        </Provider>,
    );
    fireEvent.change(getByTestId('signin-customerId'),
      { target: { value: '123', name: 'customerId' } });
    fireEvent.change(getByTestId('signin-email'),
      { target: { value: 'email@email.com', name: 'email' } });
    fireEvent.click(getByTestId('signin-signin'));
    expect(store.getState().user.isLoading).toBe(true);
  });
});
