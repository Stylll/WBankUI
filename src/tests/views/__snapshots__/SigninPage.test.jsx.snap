// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SigninPage Test should render properly 1`] = `
<body>
  <div>
    <div
      class=""
    >
      <div
        class="header container"
      >
        <div
          class="link"
        >
          WBank
        </div>
        <div
          class="topRight"
        />
      </div>
      <div
        class="bottom container"
      />
      <div
        class="signinpage container"
      >
        <div
          class="form-container"
        >
          <h2
            class="intro-text"
          >
            Welcome To WBank
          </h2>
          <div
            class="signin-container"
          >
            <div>
              <p>
                Sign in to account
              </p>
              <span
                class="error-text"
              />
              <div
                class="form-input"
              >
                <div
                  class="title"
                >
                  Enter your customer id
                </div>
                <input
                  class="text-input"
                  data-testid="signin-customerId"
                  id=""
                  name="customerId"
                  placeholder="3214"
                  type="text"
                  value=""
                />
                <span
                  class="info"
                />
                
              </div>
              <div
                class="form-input"
              >
                <div
                  class="title"
                >
                  Enter your email
                </div>
                <input
                  class="text-input"
                  data-testid="signin-email"
                  id=""
                  name="email"
                  placeholder="jimhalpert@dundermuflin.com"
                  type="text"
                  value=""
                />
                <span
                  class="info"
                />
                
              </div>
              <div
                class="btn-button"
              >
                <button
                  class="button "
                  data-testid="signin-signin"
                  type="submit"
                >
                  Signin
                </button>
              </div>
            </div>
            <br />
            <h4
              class="link"
            >
              Don't have an account?. click here to signup
            </h4>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
